packages/block-library/src/comments/index.php

```diff
@@ -0,0 +1,213 @@
+}
+
+/**
+ * Registers the `core/comments-query-loop` block on the server.
+ */
+function register_block_core_comments_query_loop() {
+	register_block_type_from_metadata(
+		__DIR__ . '/comments',
+		array(
+			'render_callback'   => 'render_block_core_comments_query_loop',
+			'skip_inner_blocks' => true,
+		)
+	);
+}
+add_action( 'init', 'register_block_core_comments_query_loop' );
+
+/**
+ * Use the button block classes for the form-submit button.
+ *
+ * @param array $fields The default comment form arguments.
+ *
+ * @return array Returns the modified fields.
+ */
+function comments_query_loop_block_form_defaults( $fields ) {
+	if ( wp_is_block_theme() ) {
+		$fields['submit_button'] = '<input name="%1$s" type="submit" id="%2$s" class="%3$s wp-block-button__link" value="%4$s" />';
+		$fields['submit_field']  = '<p class="form-submit wp-block-button">%1$s %2$s</p>';
+	}
+
+	return $fields;
+}
+add_filter( 'comment_form_defaults', 'comments_query_loop_block_form_defaults' );
+
+/**
+ * Enqueues styles from the legacy `core/post-comments` block. These styles are
+ * required by the block's fallback.
+ *
+ * @param string $block_name Name of the new block type.
+ */
+function enqueue_legacy_post_comments_block_styles( $block_name ) {
+	static $are_styles_enqueued = false;
+
+	if ( ! $are_styles_enqueued ) {
+		$handles = array(
+			'wp-block-post-comments',
+			'wp-block-buttons',
+			'wp-block-button',
+		);
+		foreach( $handles as $handle ) {
+			wp_enqueue_block_style( $block_name, array( 'handle' => $handle) );
+		}
+		$are_styles_enqueued = true;
+	}
+}
+
+/**
+ * Renders the legacy `core/post-comments` block on the server.
+ * It triggers a developer warning and then calls the renamed
+ * block's `render_callback` function output.
+ *
+ * This can be removed when WordPress X.X is released.
+ *
+ * @param array    $attributes Block attributes.
+ * @param string   $content    Block default content.
+ * @param WP_Block $block      Block instance.
+ *
+ * @return string Returns the output of the block.
+ */
+function render_legacy_post_comments_block( $attributes, $content, $block ) {
+	trigger_error(
+		/* translators: %1$s: Block type */
+		sprintf( __( 'Block %1$s has been renamed to Comments Query Loop. %1$s will be supported until WordPress version X.X.', 'gutenberg' ), $block->name ),
+		headers_sent() || WP_DEBUG ? E_USER_WARNING : E_USER_NOTICE
+	);
+	return render_block_core_comments_query_loop( $attributes, $content, $block );
+}
+
+/**
+ * Ensures backwards compatibility for any users running the Gutenberg plugin
+ * who have used Post Comments before it was merged into Comments Query Loop.
+ *
+ * The same approach was followed when core/query-loop was renamed to
+ * core/post-template.
+ *
+ * This can be removed when WordPress X.X is released.
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/41807
+ * @see https://github.com/WordPress/gutenberg/pull/32514
+ */
+function gutenberg_register_legacy_post_comments_block() {
```

Comment:
I haven't checked all the code here, but I have two notes:
1. The Query Loop rename was while the block hadn't landed in core, so I'm not sure if it should/can be the same approach. Also for me the similar case for Query Loop would be to merge with `Latest Posts` block, which we haven't.
2. Any function in the block's index.php is renamed by adding the `gutenberg_` prefix, so no functions here should have it already.

In general if something is in code only for the plugin, it should live in a different file.

I don't have all the context here, but by seeing the removed files in the legacy `post-comments` block I was wondering why the deprecation isn't handled in the block itself, set a timeline and then remove. Are there any other big technical challenges?

Reply:
The idea came from a [discussion](https://github.com/WordPress/gutenberg/issues/34994#issuecomment-1103299112) between @mtias and @bernie on the [Comments Loop Block: tracking issue](https://github.com/WordPress/gutenberg/issues/34994).

There is more info in the related issue as well: [Post Comments block: Merge Comments Query Loop block](https://github.com/WordPress/gutenberg/issues/40521#)

I think the decision of merging both blocks was more related to _replacing the block as transparently as possible_, rather than a technical challenge. 🙂

Reply:
> I don't have all the context here, but by seeing the removed files in the legacy `post-comments` block I was wondering why the deprecation isn't handled in the block itself, set a timeline and then remove.

Can you elaborate a bit on this scenario? As @DAreRodz said, we chose merging the (legacy) Post Comments block's logic into its successor block's (the Comments block) per a suggestion from @mtias -- to do it transparently, and to unclutter the block namespace, IIRC; especially since the Post Comments block _was_ indeed part of a WP release (6.0) and is used in themes (such as TT2).

However, I'd be curious if the scenario you describe would work as an alternative, and what it would look like in detail 😄